pkg/tenant/topics.go

```diff
@@ -317,46 +316,61 @@ func (s *Server) TopicDelete(c *gin.Context) {
+		if errors.Is(err, db.ErrNotFound) {
+			log.Warn().Err(err).Str("topicID", topicID.String()).Msg("topic not found")
+			c.JSON(http.StatusNotFound, api.ErrorResponse("topic not found"))
+			return
+		}
 		log.Error().Err(err).Str("topicID", topicID.String()).Msg("could not retrieve topic")
-		c.JSON(http.StatusNotFound, api.ErrorResponse("topic not found"))
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not delete topic"))
 		return
 	}
 
 	// Verify that the user owns the topic
 	if claims.OrgID != topic.OrgID.String() {
 		log.Warn().Err(err).Str("user_org", claims.OrgID).Str("topic_org", topic.OrgID.String()).Msg("topic OrgID does not match user OrgID")
-		c.JSON(http.StatusForbidden, api.ErrorResponse("user is not authorized to delete this topic"))
+		c.JSON(http.StatusNotFound, api.ErrorResponse("topic not found"))
 		return
 	}
 
 	// Send confirmation token if not provided
-	if confirm.ConfirmToken == "" {
+	if confirm.Token == "" {
 		// Create a short-lived confirmation token in the database
-		topic.ConfirmDeleteToken = secrets.CreateToken(16)
-		topic.ConfirmDeleteExpire = time.Now().Add(5 * time.Minute)
+		if topic.ConfirmDeleteToken, err = db.NewResourceToken(topic.OrgID); err != nil {
+			log.Error().Err(err).Str("topicID", topicID.String()).Msg("could not generate confirmation token")
+			c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not generate confirmation token"))
+			return
+		}
 		if err = db.UpdateTopic(ctx, topic); err != nil {
 			log.Error().Err(err).Str("topicID", topicID.String()).Msg("could not save topic")
 			c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not generate confirmation token"))
 			return
 		}
 
 		confirm.Name = topic.Name
-		confirm.ConfirmToken = topic.ConfirmDeleteToken
+		confirm.Token = topic.ConfirmDeleteToken
 		c.JSON(http.StatusOK, confirm)
 		return
 	}
 
-	// Verify that the confirmation token is correct
-	if confirm.ConfirmToken != topic.ConfirmDeleteToken {
-		log.Warn().Err(err).Msg("confirmation token does not match")
-		c.JSON(http.StatusBadRequest, api.ErrorResponse("unrecognized confirmation token"))
+	// Verify the confirmation token is valid and not expired
+	token := &db.ResourceToken{}
+	if err = token.Decode(confirm.Token); err != nil {
+		log.Warn().Err(err).Msg("could not decode confirmation token")
+		c.JSON(http.StatusPreconditionFailed, api.ErrorResponse("invalid confirmation token"))
+		return
+	}
+
+	if token.IsExpired() {
+		log.Warn().Msg("confirmation token has expired")
+		c.JSON(http.StatusPreconditionFailed, api.ErrorResponse("invalid confirmation token"))
 		return
 	}
 
-	// Verify the confirmation token has not expired
-	if time.Now().After(topic.ConfirmDeleteExpire) {
-		log.Warn().Err(err).Msg("confirmation token has expired")
-		c.JSON(http.StatusBadRequest, api.ErrorResponse("confirmation token has expired"))
+	// Verify that we have the right token
+	if token.ID.Compare(topic.ID) != 0 {
```

Comment:
Am I missing something? It looks like the OrgID is put in but then it is compared to the topic ID. If I am not missing anything why aren't the tests catching this?

Reply:
My bad, it should be the topic ID and the test coverage isn't complete yet so I didn't catch it.

Reply:
No worries - that's what code reviews are for!